### DISCUSSION TITLE
How can I calculate the relative circumference of the points of a profile or airfoil?

### DISCUSSION THREAD

**Discussion by Mtabei** [DLR-SL/CPACS#739] [Q&A] [answered]
How can I calculate the relative circumference of the points of a profile or airfoil?
I tried fitting the points to a circle for fuselage profiles but the results were not very accurate, and for the airfoils I can only identify the leading and trailing edge points.

**Comment by MarAlder** [comment, marked as answer]
You have a CPACS file with guide curves, I guess? Then I suggest to use TiGL's configuration manager API.

The following example provides the guide curve points of the [simpleAircraft.xml](https://github.com/DLR-SL/CPACS/blob/develop/examples/simpleAircraft.xml) example:

```python 
from tixi3 import tixi3wrapper
from tigl3 import tigl3wrapper
import tigl3.configuration
import numpy as np

# Instantiate TiXI
tixi_h = tixi3wrapper.Tixi3()

# Open the XML file
fname = 'simpleAircraft.xml'
error = tixi_h.open(fname)
if not error:
    print('CPACS data set %s opended successfully.'%fname)

tigl_h = tigl3wrapper.Tigl3()
tigl_h.open(tixi_h,'aircraftModel')

# Load config manager
mgr = tigl3.configuration.CCPACSConfigurationManager_get_instance()
aircraft_config = mgr.get_configuration(tigl_h._handle.value)

# Fuselage; replace "get_fuselage(1)" with "get_wing(2)" to geht the vertical tailplane with guide curves from the given example
fuselage = aircraft_config.get_fuselage(1)
gc_points = fuselage.get_guide_curve_points()
gc_coords = np.zeros([len(gc_points),3])
for i,pnt in enumerate(gc_points):
    gc_coords[i] = pnt.Coord()

gc_wires = fuselage.get_guide_curve_wires() # If you would like to do further CAD operations, this might be a good starting point

**Reply by Mtabei** [reply]
Unfortunately I am working with C# so I will not be able to use the API for this. do you have any other recommendation?

**Reply by MarAlder** [reply]
Puh... so you're not using the TiGL library (C, C++, Python, Java, MATLAB, and FORTRAN) and you don't have a lofting algorithm at hand? You might derive the transformed profile coordinates at the section elements from the various transformations and then spline the profile. But I guess you also want to compute the coordinates at the base points of the guide curve in between the sections, right?

**Reply by Mtabei** [reply]
Yes that's what I am trying exactly. I have the profile points and I need to identify the section point where I should start the guide curve

**Reply by MarAlder** [reply]
Any progress with your C# implementation? Maybe @rainman110 or @joergbrech can give some methodological tips on the general approach?

**Reply by rainman110** [reply]
It's unfortunately more complicated than it sounds. First let me try to explain the general math / method (i.e. without using TiGL):

1. Create a B-spline through all profiles
2. Project the point onto the B-spline to get the curve parameter u
3. Determine the curve curve length from start to parameter u by integrating along the curve's path: https://en.wikipedia.org/wiki/Differentiable_curve#Length_and_natural_parametrization
4. Determine the total length of the curve using the same algorithm

Fortunately TiGL and OpenCASCADE has several methods that help here
1. You can directly query the b-spline wire from tigl instead of creating it
2. Use TiGL's function ``double ProjectPointOnWire(const TopoDS_Wire& wire, gp_Pnt p);`` to get the relative circumference of the point p ont the wire

**Reply by Mtabei** [reply]
Thank you so much, this helps a lot
